[JumpThreading] Remove deleted BB from Unreachable

weiguozhi opened the pull request:
Although an unreachable BB is skipped by processBlock, its successor can still be handled by processBlock, and maybeMergeBasicBlockIntoOnlyPred may merge the two BBs and delete the unreachable BB. Then the garbage pointer is left in Unreachable set. This patch removes the invalid BB pointer from the Unreachable set.

weiguozhi committed:
[JumpThreading] Remove deleted BB from Unreachable

Although an unreachable BB is skipped by processBlock, its successor can
still be handled by processBlock, and maybeMergeBasicBlockIntoOnlyPred
may merge the two BBs and delete the unreachable BB. Then the garbage
pointer is left in Unreachable set. This patch removes the invalid BB
pointer from the Unreachable set.

kazutakahirata commented:
> Although an unreachable BB is skipped by processBlock, its successor can still be handled by processBlock, and maybeMergeBasicBlockIntoOnlyPred may merge the two BBs and delete the unreachable BB. Then the garbage pointer is left in Unreachable set. This patch removes the invalid BB pointer from the Unreachable set.

Ugh.  I've encountered essentially the same problem elsewhere in LLVM (although I don't remember where).  Meanwhile, in the inliner, when all call sites of a given `static` function have been inlined, we add the function to the list of dead functions, and then remove those dead functions after we are done inlining.

```
  if (Unreachable.count(SinglePred) && DTU->isBBPendingDeletion(SinglePred))
    Unreachable.erase(SinglePred);
```

I have a few questions about this:

- Is this the only place where a live basic block appears to be "Unreachable"?
- If `DTU->isBBPendingDeletion(SinglePred)` above is false, are we OK?  Does somebody else remove `SinglePred` from `Unrechable` for us?
- Is it feasible to keep all basic blocks without erasing them?  I'm guessing that this might be difficult because we sometimes call external utility functions to manipulate basic blocks, and those utility functions might remove basic blocks.

Thanks!

nikic reviewed (commented):
Can we bail out of maybeMergeBasicBlockIntoOnlyPred on unreachable pred instead?

weiguozhi commented:
> ```
>   if (Unreachable.count(SinglePred) && DTU->isBBPendingDeletion(SinglePred))
>     Unreachable.erase(SinglePred);
> ```
> 
> I have a few questions about this:
> 
> * Is this the only place where a live basic block appears to be "Unreachable"?

A live basic block in Unreachable is not a problem. A deleted basic block in Unreachable is a problem. Here the SinglePred has not been deleted, but it is marked so in DTU. It will be actually deleted when we call DTU->flush() later.

> * If `DTU->isBBPendingDeletion(SinglePred)` above is false, are we OK?  Does somebody else remove `SinglePred` from `Unrechable` for us?

`DTU->isBBPendingDeletion(SinglePred)` is false means SinglePred will not be deleted, so it is OK for us. Current implementation never delete any pointer from `Unrechable`.

> * Is it feasible to keep all basic blocks without erasing them?  I'm guessing that this might be difficult because we sometimes call external utility functions to manipulate basic blocks, and those utility functions might remove basic blocks.

Maybe @nikic's suggestion can accomplish this. I will try it.

weiguozhi committed:
[JumpThreading] Bail out maybeMergeBasicBlockIntoOnlyPred if SinglePred is in Unreachable.

Merge SinglePred and its successor may delete SinglePred, and causes
garbage pointer in Unreachable. Bail out
maybeMergeBasicBlockIntoOnlyPred if we found this case.

weiguozhi commented:
> Can we bail out of maybeMergeBasicBlockIntoOnlyPred on unreachable pred instead?

Done. But it causes regression in pr62908.ll.

weiguozhi commented:
ping

nikic commented on llvm/test/Transforms/JumpThreading/pr62908.ll:
Can you please regenerate the test so the whole IR is visible?

weiguozhi replied:
done.

nikic reviewed (commented):
Looking at this again, I'm not sure I fully understand what problem you're trying to solve here.

This code uses DTU so block deletion is deferred, so there are no problems with address reuse. Why is it problematic to merge into an unreachable block?

weiguozhi committed:
Regenerate the test result so the whole IR is included.

weiguozhi commented:
> Looking at this again, I'm not sure I fully understand what problem you're trying to solve here.
> 
> This code uses DTU so block deletion is deferred, so there are no problems with address reuse. Why is it problematic to merge into an unreachable block?

As you have noticed MergeBasicBlockIntoOnlyPred can mark an unreachable block for deletion. The deletion is deferred until 
DTU->flush() is called, then the unreachable BB is actually deleted, but the set Unreachable is not modified, so a deleted BB pointer is left in Unreachable. 

In later optimizations new basic blocks can be generated, if we are unlucky a new BB may be created at the deleted BB address, then the new BB pointer is automatically contained in Unreachable unexpectedly. And the new BB will be skipped by the following optimization. This part is a random behavior, so we got indeterministic result on our real code.

nikic commented:
> As you have noticed MergeBasicBlockIntoOnlyPred can mark an unreachable block for deletion. The deletion is deferred until DTU->flush() is called, then the unreachable BB is actually deleted, but the set Unreachable is not modified, so a deleted BB pointer is left in Unreachable.

I think the last bit I'm still missing is when JumpThreading calls DTU->flush() *during* the transform. I thought JT only flushes DT updates after everything is done, at which point the Unreachable set is not relevant anymore.

weiguozhi commented:
> > As you have noticed MergeBasicBlockIntoOnlyPred can mark an unreachable block for deletion. The deletion is deferred until DTU->flush() is called, then the unreachable BB is actually deleted, but the set Unreachable is not modified, so a deleted BB pointer is left in Unreachable.
> 
> I think the last bit I'm still missing is when JumpThreading calls DTU->flush() _during_ the transform. I thought JT only flushes DT updates after everything is done, at which point the Unreachable set is not relevant anymore.

DTU->flush() is called in the chain:
```
    threadThroughTwoBasicBlocks()
->  getOrCreateBFI()
->  runExternalAnalysis<BlockFrequencyAnalysis>()
->  DTU->flush()

```

nikic reviewed (approved):
LGTM, thanks for the explanation. I think what JumpThreading does with BPI/BFI is unnecessarily complicated (why do we ever compute a fresh BPI/BFI if we're only doing updates?) and could maybe be cleaned up to avoid this flush, but let's land this fix for now.
